R/rvar-.R

```diff
@@ -0,0 +1,1066 @@
+  # TODO: probably could do something more efficient here and for restore
+  .draws = draws_of(x)
+  out <- vec_chop(aperm(.draws, c(2, 1, seq_along(dim(.draws))[c(-1,-2)])))
+  for (i in seq_along(out)) {
+    attr(out[[i]], "nchains") <- nchains(x)
+  }
+  out
+}
+
+#' @importFrom vctrs vec_restore
+#' @export
+vec_restore.rvar <- function(x, ...) {
+  if (length(x) > 0) {
+    # need to handle the case of creating NAs from NULL entries so that
+    # vec_init() works properly: vec_init requires vec_slice(x, NA_integer_)
+    # to give you back NA values, but this breaks because we use lists as proxies.
+    # When using a list as a proxy, a proxy entry in `x` that is equal to NULL
+    # actually corresponds to an NA value due to the way that list indexing
+    # works: when you do something like list()[c(NA_integer_,NA_integer_)]
+    # you get back list(NULL, NULL), but when you do something like
+    # double()[c(NA_integer_,NA_integer_)] you get back c(NA, NA).
+    # So we have to make the NULL values be NA values to mimic vector indexing.
+
+    # N.B. could potentially do this with vec_cast as well (as long as the first
+    # dimension is the slicing index)
+    x[sapply(x, is.null)] <- list(array(NA, dim = c(1,1)))
+
+  }
+  # broadcast dimensions and bind together
+  new_dim <- dim_common(lapply(x, dim))
+  .draws <- abind(lapply(x, broadcast_array, new_dim), along = 1)
+  # move draws dimension back to the front
+  if (!is.null(.draws)) {
+    .draws <- aperm(.draws, c(2, 1, seq_along(dim(.draws))[c(-1,-2)]))
+  }
+  # determine the number of chains
+  nchains_or_null <- lapply(x, function(x) if (dim(x)[[2]] %||% 1 == 1) NULL else attr(x, "nchains"))
+  .nchains <- Reduce(nchains2_common, nchains_or_null) %||% 1L
+
+  new_rvar(.draws, .nchains = .nchains)
+}
+
+
+# distributional stuff ----------------------------------------------------
+
+#' Density, CDF, and quantile functions of random variables
+#'
+#' The probability density function (`density()`), cumulative distribution
+#' function (`cdf()`), and quantile function / inverse CDF (`quantile()`) of
+#' an [`rvar`].
+#'
+#' @param x an [`rvar`]
+#' @param q,at vector of quantiles.
+#' @param probs vector of probabilities
+#' @param ... Additional arguments passed onto underlying methods:
+#'   - For `density()`, these are passed to [stats::density()].
+#'   - For `cdf()`, these are ignored.
+#'   - For `quantile()`, these are passed to [stats::quantile()].
+#'
+#' @return
+#'
+#' A vector of the same length as the input (`q`, `at`, or `probs`) containing
+#' values from the corresponding function of the given [`rvar`].
+#'
+#' @name rvar-functions
+#' @export
+density.rvar <- function(x, at, ...) {
+  if (length(x) != 1) {
+    stop2("density() can currently only be used on scalar rvars")
+  }
+
+  d <- density(draws_of(x), cut = 0, ...)
+  f <- approxfun(d$x, d$y, yleft = 0, yright = 0)
+  f(at)
+}
+
+#' @importFrom distributional cdf
+#' @export
+distributional::cdf
```

Comment:
I had copied this style after [this stackoverflow issue](https://stackoverflow.com/a/39621289) and [usage in tidyverse packages](https://github.com/tidyverse/dplyr/blob/master/R/reexport-tibble.r), which generates a `reexport.Rd` file with links to re-exported functions. But I tried what you suggested and it seems to work better (it doesn't generate the redundant reexports.Rd file but everything else --- links etc --- seems to still work). So I've switched to that. Perhaps that style was not intended for use with S3 generics...

Reply:
nevermind --- I tried the method you suggested and I got warnings about undocumented code objects, which perhaps explains why the approach I was using is recommended. I have reverted to the approach I used about and I am unresolving this conversation in case there is something I misunderstood about your suggestion.